### PR TITLE
Expose `log-file` mount option for GlusterFS

### PR DESCRIPTION
**Pull request by bodgit**
GlusterFS by default uses a log file based on the mountpoint path munged into a
file, i.e. `/mnt/foo/bar` becomes `/var/log/glusterfs/mnt-foo-bar.log`.

On certain Kubernetes environments this can result in a log file that exceeds
the 255 character length most filesystems impose on filenames causing the mount
to fail. This option exists to set the log file to something sane.

I've tested this with the following simple pod:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: glusterfs
  namespace: default
spec:
  containers:
  - image: busybox
    command:
    - sleep
    - "3600"
    name: glusterfs
    volumeMounts:
    - mountPath: /mnt/glusterfs
      name: glusterfsvol
  volumes:
  - name: glusterfsvol
    glusterfs:
      endpoints: glusterfs-cluster
      path: gv0
      #logFile: "/var/log/glusterfs/test.log"
      readOnly: true
  restartPolicy: Always
```

(Assuming the glusterfs-cluster endpoint is created as per the documentation)

Without the `logFile` option, I get the log file `/var/lib/glusterfs/var-lib-kubelet-pods-4cc78b01-57cd-11e5-96eb-080027fdddda-volumes-kubernetes.io~glusterfs-glusterfsvol.log` created. Uncommenting the `logFile` option gives me a `/var/log/glusterfs/test.log` file as expected.

I wasn't entirely sure how to handle the Persistent Volumes in `getGlusterVolumeSource()` so for now I just returned an empty string, suggestions welcome.

This should fix the problem I observed in #12575

### CODE REVIEW

**Comment by rootfs** [comment]
exposing log file as a user visible parameter is not ideal, one argument is that you have to know the path to the log file.

Would it be possible to just reuse your code in glusterfs.go but make log-file to live inside kubelet pod directory and remove the api level change?

**Comment by bodgit** [comment]
Sure, I think so. How do I retrieve the kubelet pod directory?

**Comment by rootfs** [comment]
@bodgit the glusterfs log file name is based on mountpoint, so I think if you play some tricks like base64 encoding the pod UID [around this line](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L171), you probably are able to reduce the filename length without setting log-file option.

that line also points to the pod volume directory.

**Comment by rootfs** [comment]
@bodgit can you sign cla please?

**Comment by bodgit** [comment]
I'm not sure I see how base64 encoding anything helps. Without setting the `log-file` mount option, glusterfs will try and create a log file based on where kubelet is trying to mount it.

The problem is that when running Kubernetes on top of Mesos, the path to the top pod directory is something like `/tmp/mesos/slaves/20150804-092332-1680016394-5050-1110-S0/frameworks/20150715-130216-1713570826-5050-1229-0000/executors/e3df79b24ad1d75b_k8sm-executor/runs/cf292370-4ee6-4fa1-8e38-51a980db170e/pods/4c4faec6-57a1-11e5-8e05-005056b10d4c` which is already ~ 230 characters and near the filename length limit as opposed to running Kubernetes on its own where the path to a pod directory is something like `/var/lib/kubelet/pods/4cc78b01-57cd-11e5-96eb-080027fdddda` which is much shorter. Adding the extra path info to where the volume is actually mounted sends us over the 255 character limit.

I'm trying to find an API call to return `/var/lib/kubelet/pods/<pod UID>` which looks like the right place to write the logs. I can see there's a `getPodDir(podUID)` call in `pkg/kubelet/kubelet.go` but I can't work out how I can call that. Until today I hadn't written any Go so I'm not very familiar with the language :wink:.

**Comment by rootfs** [comment]
ok, thanks for digging it out. 

you can see if this patch works. The idea is based on our discussion to move gluster log file to pod volume directory. This comes with a downside though, if there is something wrong the pod, we don't have any glusterfs log to help us to figure out what goes wrong.

```diff
diff --git a/pkg/volume/glusterfs/glusterfs.go b/pkg/volume/glusterfs/glusterfs.go
index bf9efda..12ca0c3 100644
--- a/pkg/volume/glusterfs/glusterfs.go
+++ b/pkg/volume/glusterfs/glusterfs.go
@@ -19,6 +19,7 @@ package glusterfs
 import (
        "math/rand"
        "os"
+       "path"
 
        "github.com/golang/glog"
        "k8s.io/kubernetes/pkg/api"
@@ -219,6 +220,8 @@ func (b *glusterfsBuilder) setUpAtInternal(dir string) error {
        if b.readOnly {
                options = append(options, "ro")
        }
+       log := path.Join(dir, "..", b.glusterfs.volName+".log")
+       options = append(options, "log-file="+log)
 
        l := len(b.hosts.Subsets)
        // Avoid mount storm, pick a host randomly.

```

Alternatively, you can also make use of the plugin directory like `/var/lib/kubelet/plugins/kubernetes.io/glusterfs/10.16.154.81/glusterfsvol/glusterfs.log`, which is more persistent. The patch is as the following.

```diff
diff --git a/pkg/volume/glusterfs/glusterfs.go b/pkg/volume/glusterfs/glusterfs.go
index bf9efda..ae0f10c 100644
--- a/pkg/volume/glusterfs/glusterfs.go
+++ b/pkg/volume/glusterfs/glusterfs.go
@@ -19,6 +19,7 @@ package glusterfs
 import (
        "math/rand"
        "os"
+       "path"
 
        "github.com/golang/glog"
        "k8s.io/kubernetes/pkg/api"
@@ -226,10 +227,17 @@ func (b *glusterfsBuilder) setUpAtInternal(dir string) error {
        // Iterate all hosts until mount succeeds.
        for i := start; i < start+l; i++ {
                hostIP := b.hosts.Subsets[i%l].Addresses[0].IP
+               p := path.Join(b.glusterfs.plugin.host.GetPluginDir(glusterfsPluginName), hostIP, b.glusterfs.volName)
+               if err := os.MkdirAll(p, 0750); err == nil {
+                       log := path.Join(p, "glusterfs.log")
+                       options = append(options, "log-file="+log)
+               }
+
                errs = b.mounter.Mount(hostIP+":"+b.path, dir, "glusterfs", options)
                if errs == nil {
                        return nil
                }
+               os.RemoveAll(p)
        }
        glog.Errorf("Glusterfs: mount failed: %v", errs)
        return errs

```

**Comment by bodgit** [comment]
@rootfs I based my change on your second version, however I've just created a directory without the gluster server IP as it shouldn't matter which host the mount succeeds on. Also if the mount fails for the first server then I think you'll end up with multiple `log-file` mount options in the array. I also made creating the directory for the log file mandatory and then changed to use `os.MkdirAll` as it was failing and then resorting back to using the default log file. I've tested this on my cluster and it seems to work.

**Comment by rootfs** [comment]
@bodgit thanks for putting it together. The idea of using host IP in the directory is that if there are multiple glusterfs clusters and somehow they use the same gluster volume name, using host IP in the directory can then be able to distinguish the clusters. Since this directory is removed if mount fails, no residual log files will be left.

File: pkg/volume/glusterfs/glusterfs.go [hunk 1]
```diff
@@ -221,6 +222,13 @@ func (b *glusterfsBuilder) setUpAtInternal(dir string) error {
 		options = append(options, "ro")
 	}
 
+	p := path.Join(b.glusterfs.plugin.host.GetPluginDir(glusterfsPluginName), b.glusterfs.volName)
+	if err := os.MkdirAll(p, 0750); err != nil {
+		return err
```
Comment (rootfs) [inline pkg/volume/glusterfs/glusterfs.go]: this is rare, but I prefer not exiting so early. How about if `MkdirAll` succeeds, then append `log-file` option.

**Comment by rootfs** [comment]
@bodgit just a nit, then LGTM.

File: pkg/volume/glusterfs/glusterfs.go (hunk 1 above)
Comment (bodgit) [inline pkg/volume/glusterfs/glusterfs.go]: I had mistakenly used `os.Mkdir` initially which was failing due to parent directories being missing. The code then silently continued without the mount option and defaulted back to the original behaviour which I was trying to track down so I changed the logic which then made my error obvious.

Comment (rootfs) [inline pkg/volume/glusterfs/glusterfs.go]: that's a nice debug trick :)

Comment (rootfs) [inline pkg/volume/glusterfs/glusterfs.go]: @bodgit have you signed cla? The label is still red

**Comment by bodgit** [comment]
I'm testing with a GlusterFS cluster consisting of three nodes, I've shut down the first two to test the directories being created and deleted however kubelet seems to be sticking to the first node and not trying the others listed in the endpoint.

My endpoint is defined as follows:

```yaml
kind: "Endpoints"
apiVersion: "v1"
metadata: 
  name: "glusterfs-cluster"
subsets: 
  - addresses: 
      - ip: "10.245.1.11"
    ports: 
      - port: 1
  - addresses: 
      - ip: "10.245.1.12"
    ports: 
      - port: 1
  - addresses: 
      - ip: "10.245.1.13"
    ports: 
      - port: 1
```

10.245.1.{11,12} are stopped so only 10.245.1.13 is up. I've added debug and the problem is `l := len(b.hosts.Subsets)` is returning 1, not 3 which means `start` and `i` are always 0 hence the first server is the only one ever polled and no others are tried. Any ideas @rootfs?

**Comment by rootfs** [comment]
@bodgit open a new issue - endpoints api underwent several changes and makes you the first victim.

**Comment by bodgit** [comment]
I think I've spotted the problem. The number of subsets in the endpoint is always 1 because the port is the same for each address, so actually what you get is one subset with 3 addresses. I think if you were to use three different ports then you would end up with 3 subsets each with 1 address each.

I made the following quick change:

```diff
diff --git a/pkg/volume/glusterfs/glusterfs.go b/pkg/volume/glusterfs/glusterfs.go
index 8981e16..1fed162 100644
--- a/pkg/volume/glusterfs/glusterfs.go
+++ b/pkg/volume/glusterfs/glusterfs.go
@@ -229,12 +229,12 @@ func (b *glusterfsBuilder) setUpAtInternal(dir string) error {
        log := path.Join(p, "glusterfs.log")
        options = append(options, "log-file="+log)
 
-       l := len(b.hosts.Subsets)
+       l := len(b.hosts.Subsets[0].Addresses)
        // Avoid mount storm, pick a host randomly.
        start := rand.Int() % l
        // Iterate all hosts until mount succeeds.
        for i := start; i < start+l; i++ {
-               hostIP := b.hosts.Subsets[i%l].Addresses[0].IP
+               hostIP := b.hosts.Subsets[0].Addresses[i%l].IP
                errs = b.mounter.Mount(hostIP+":"+b.path, dir, "glusterfs", options)
                if errs == nil {
                        return nil
```

And now the Pod comes up and uses the last node in the cluster after potentially trying the first two which are down.

**Comment by rootfs** [comment]
cool, a more comprehensive fix is to iterate all Subsets and their Addresses, so we can deal with different Endpoints formation.

**Comment by rootfs** [comment]
@bodgit here is my fix, it *seems* to work for both endpoints formation.

```diff
diff --git a/pkg/volume/glusterfs/glusterfs.go b/pkg/volume/glusterfs/glusterfs.go
index 449c9e9..4866512 100644
--- a/pkg/volume/glusterfs/glusterfs.go
+++ b/pkg/volume/glusterfs/glusterfs.go
@@ -226,10 +226,14 @@ func (b *glusterfsBuilder) setUpAtInternal(dir string) error {
        start := rand.Int() % l
        // Iterate all hosts until mount succeeds.
        for i := start; i < start+l; i++ {
-               hostIP := b.hosts.Subsets[i%l].Addresses[0].IP
-               errs = b.mounter.Mount(hostIP+":"+b.path, dir, "glusterfs", options)
-               if errs == nil {
-                       return nil
+               al := len(b.hosts.Subsets[i%l].Addresses)
+               astart := rand.Int() % al
+               for j := astart; j < astart+al; j++ {
+                       hostIP := b.hosts.Subsets[i%l].Addresses[j%al].IP
+                       errs = b.mounter.Mount(hostIP+":"+b.path, dir, "glusterfs", options)
+                       if errs == nil {
+                               return nil
+                       }
                }
        }
        glog.Errorf("Glusterfs: mount failed: %v", errs)
```

**Comment by bodgit** [comment]
I fixed it in a slightly different way by building a map which should mean there's no duplicate IP addresses and as I understand it the order of the keys is random so the loop over each IP is simpler as you just go from beginning to end.

**Comment by rootfs** [comment]
map is good idea. but i'd love to keep the randomizer in place. this is useful to avoid mount storm where you have tons of pods started simultaneously and all want to mount from the same gluster host.

**Comment by bodgit** [comment]
According to http://blog.golang.org/go-maps-in-action#TOC_7 the order is not defined between iterations so it should still have the same behaviour.

**Comment by rootfs** [comment]
just tested it, there seems to be some randomness.

**Comment by rootfs** [comment]
@bodgit please check cla.

@mikedanese once cla label cleared, then LGTM.

**Comment by bodgit** [comment]
CLA signed

**Pull request merged**